### ISSUE TITLE
XLA:CPU `__ynn_fusion` (slinky) crashes or hangs on rank-7/8 broadcast-multiply-reduce — regression in jaxlib 0.10.2 & 0.11.0 on Linux x86-64

### ISSUE THREAD

**Issue by conorheins** [jax-ml/jax#39257]
## Description

Starting with jaxlib **0.10.2** (still present in **0.11.0**), this 10-line CPU program kills the process on Linux x86-64 — SIGSEGV on some CPUs, an apparent infinite hang on others:

```python
import jax
import jax.numpy as jnp

def f(obs, A):
    lik = jnp.sum(obs[..., None, None, None, None, None] * A, axis=2)   # -> (10,4,9,9,9,9,9)
    return jnp.log(jnp.maximum(lik, 2.220446049250313e-16))

obs = jnp.ones((10, 4, 6), jnp.float32)
A = jnp.ones((10, 4, 6, 9, 9, 9, 9, 9), jnp.float32)
r = jax.jit(f)(obs, A)   # <- SIGSEGV / hang on jaxlib 0.10.2 & 0.11.0, Linux x86-64
print(r.shape)
```

A broadcast of a rank-3 operand into a rank-8 shape, elementwise multiply, one-axis reduce, log — ~1.5 MB of data. On jaxlib 0.10.1 it runs in well under a second, as it does on Linux aarch64 and macOS arm64 with **all** versions including 0.10.2/0.11.0.

## Crash backtrace (bare metal, AMD EPYC 9554, jaxlib 0.11.0, Python 3.13)

```
Thread 62 "tf_XLAEigen" received signal SIGSEGV, Segmentation fault.
#0  int slinky::optimize_dims<slinky::buffer<void const, 8ul> >(slinky::raw_buffer&, slinky::buffer<void const, 8ul>&) ()
    from .../site-packages/jaxlib/libjax_common.so
#1  std::_Function_handler<long (slinky::call_stmt const*, slinky::eval_context&),
    slinky::func::make<ynn::(anonymous namespace)::make_transpose_impl(int, std::vector<int, std::allocator<int> >)::$_0>(...)::{lambda(...)#1}>::_M_invoke(...) ()
    from .../site-packages/jaxlib/libjax_common.so
#2  0xffffffffffffffff in ?? ()
```

i.e. the crash is inside the **`ynn` fusion's slinky pipeline** — `slinky::optimize_dims` on a `slinky::buffer<void const, 8>` created by `ynn::make_transpose_impl` — on the `tf_XLAEigen` runtime thread, so the module compiles and dies when the fused kernel first runs. Consistent with that, the HLO side looks healthy: with `--xla_dump_to`, all optimization passes complete, buffer assignment and the memory-usage report are dumped, and the entry root of the optimized module is:

```
ROOT %ynn_fusion = f32[...] fusion(%A.1, %x.1), kind=kCustom, calls=%fused_computation,
    backend_config={"outer_dimension_partitions":[],"fusion_config":{"kind":"__ynn_fusion"}}
```

where `%fused_computation` is just `broadcast → multiply → reduce`.

Disabling the ynn fusion path fixes it. `XLA_FLAGS=--xla_cpu_experimental_ynn_fusion_type=` (empty value) makes every failing configuration below compile and run instantly on both 0.10.2 and 0.11.0. Neither `--xla_cpu_use_xnnpack=false` nor `--xla_cpu_experimental_xnn_graph_fusion_mode=XNN_GRAPH_FUSION_MODE_DISABLED` helps. The regression window matches this path being enabled by default between the 0.10.1 and 0.10.2 XLA pins.

## The failure threshold moves with the CPU (all Linux x86-64, PyPI wheels)

The same bug class triggers at different ranks on different microarchitectures, and manifests as a hang on some and SIGSEGV on others:

| hardware | rank-7 variant¹ | rank-8 program (above) |
|---|---|---|
| AMD EPYC 9554 (Zen4, AVX-512) — bare metal | compiles fine | **SIGSEGV** (backtrace above) |
| GitHub Actions `ubuntu-latest` (AVX2-class) | **hang >2000 s**, uninterruptible, deterministic across Py 3.11–3.14 and re-runs | **hang >2000 s** (Py 3.12 job reached it after rank-7 passed there) |
| QEMU 9.2 x86-64 emulation (AVX2, no AVX-512) | **SIGSEGV** | **SIGSEGV** |
| Linux aarch64 / macOS arm64 (any of the above programs) | fine | fine |

Here's the rank-7 variant, it's same shape of computation but one rank lower, hangs GitHub runners on 0.10.2/0.11.0 and is where we first hit this in CI:

```python
def f(x, A):
    return jnp.sum(x[:, :, None, None, None, None, None] * A, axis=1)

x = jnp.ones((6, 3), dtype=jnp.float32)
A = jnp.ones((6, 3, 3, 3, 3, 3, 3), dtype=jnp.float32)
```

Boundary notes: rank 6 is always fine; the broadcast is required (full-shape `B * A` multiply passes); the same contraction written as `jnp.einsum` passes (lowers to dot, not a ynn elementwise fusion); reduce-axis position doesn't matter; all-uniform tiny dims (all 2s or all 3s at rank 7) escape fusion selection and pass.

First hit in a GH project's CI, where rank-7/8 programs of this shape hang every run at nbval's 2000 s timeout (fail: https://github.com/infer-actively/pymdp/actions/runs/29545844234, https://github.com/infer-actively/pymdp/actions/runs/28756780890; green-on-0.10.1 control: https://github.com/infer-actively/pymdp/actions/runs/28938528931). Happy to provide HLO dumps, larger reproducers, or run further experiments on the affected hardware.

### System info (python version, jaxlib version, accelerator, etc.)

`jax.print_environment_info()` from the failing environment (also fails identically: bare-metal RHEL 9.6 / Python 3.13.5 above; GitHub Actions ubuntu-24.04 / Python 3.11–3.14):

```
jax:    0.11.0
jaxlib: 0.11.0
numpy:  2.5.1
python: 3.12.13 (main, Jul 14 2026, 02:09:00) [GCC 14.2.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='b0e36451300e', release='6.8.0-117-generic', version='#117-Ubuntu SMP PREEMPT_DYNAMIC Thu May  7 17:26:37 UTC 2026', machine='x86_64')
```

CPU backend only. Also reproduced with jax/jaxlib 0.10.2 (Python 3.11–3.14, numpy 2.4.x).

**Comment by hawkinsp** [comment]
@penpornk

**Comment by penpornk** [comment]
@dsharletg

**Comment by rogeriojorge** [comment]
Hit this independently from a gyrokinetics code and arrived at the same place; adding a second data point plus a sharper trigger, and a note on where the fix already is.

### The fix is already merged in XNNPACK and already in XLA main

google/XNNPACK@53a1797ba ("Fix crash due to not allowing internal extra dimensions", 2026-07-17) changes exactly the type in the backtrace here:

```diff
-    slinky::buffer<void, YNN_MAX_TENSOR_RANK> sliced_output = output;
-    slinky::buffer<const void, YNN_MAX_TENSOR_RANK> sliced_input;
+    slinky::buffer<void, max_tensor_rank> sliced_output = output;
+    slinky::buffer<const void, max_tensor_rank> sliced_input;
```

with `max_tensor_rank = YNN_MAX_TENSOR_RANK + internal_extra_dims` (8 → 10). That is the `slinky::buffer<void const, 8ul>` in both of our stacks.

XLA main's XNNPACK pin is `8388bd78690515166d59f1b28e593a455a41d580` (2026-08-11), which `compare` reports as 148 commits **ahead of** and 0 behind `53a1797ba` — so current XLA already carries it, and this should be gone in the next jaxlib built from it. jaxlib 0.10.2 was uploaded 2026-06-17, a month before the fix landed.

### Why 0.10.2 specifically

It is the `sum(a*b)` → `dot` rewrite reaching ynn, matching that commit message. Counting `__ynn_fusion` in the optimized HLO for the same program on the same machine:

```
jaxlib 0.10.1: ynn_fusion ops = 0   (runs fine)
jaxlib 0.10.2: ynn_fusion ops = 1   (SIGSEGV)
```

### Smaller reproducer

No broadcast and no `log` needed — four lines:

```python
import jax, jax.numpy as jnp
S = (3, 1, 2, 2, 8, 8, 8)
u = jax.random.normal(jax.random.PRNGKey(0), S, dtype=jnp.float32)
v = jax.random.normal(jax.random.PRNGKey(1), S, dtype=jnp.float32)
jax.jit(lambda a, b: jnp.sum(a * b, axis=3))(u, v)   # SIGSEGV
```

### The trigger is a degenerate dim that is not the reduced axis

| case | result |
|---|---|
| `(3,1,2,2,8,8,8)`, reduce axis 3 | **SIGSEGV** |
| `(1,3,2,2,8,8,8)`, reduce axis 3 | **SIGSEGV** |
| `(3,2,2,2,8,8,8)`, reduce axis 3 — same rank, no size-1 dim | ok |
| `(3,1,2,2,8,8,8)`, reduce axis 1 — the size-1 dim *is* the reduced axis | ok |
| `(3,2,2,8,8,8)`, reduce axis 2 — rank 6 | ok |
| `(2,1,2,2,2,2,2)`, reduce axis 3 — rank 7 but small | ok (not routed to ynn) |
| any of the above in **float64** | ok (ynn path is f32-only) |
| CUDA backend | ok |

So it needs rank ≥ 7, a size-1 dimension other than the reduced one, f32, and enough work for the cost model to pick ynn.

### Practical workaround

`JAX_ENABLE_X64=1` avoids it entirely, because the ynn path is f32-only. That turned out to be why our CI never saw this while local runs died — every CI job exported it.

No XLA flag helps: `--xla_cpu_use_xnnpack=false`, `--xla_cpu_use_fusion_emitters=false`, `--xla_cpu_experimental_ynn_fusion_type`, `--xla_cpu_experimental_xnn_graph_fusion_mode`, `--xla_cpu_max_isa=AVX2` and `--xla_cpu_parallel_codegen_split_count=1` all leave the `__ynn_fusion` ops in the optimized HLO and still crash.

### Second machine confirmation

Intel Xeon W-2295 (AVX-512, 36 cores), Linux x86-64, Python 3.11.15, numpy 2.4.6, jaxlib 0.10.2 — SIGSEGV rather than hang. Fault address `0x3ffc7`, far from the faulting thread's stack region, and raising `ulimit -s` to 512 MB or running from a 512 MB-stack thread does not help, confirming it is a bad pointer rather than stack exhaustion.

One thing worth flagging for anyone else landing here: the fault is on an Eigen worker thread, so the Python-level traceback lands wherever the main thread happened to be. Across runs of the same program I saw it blamed on `array.py::_value`, on `compiler.py::backend_compile_and_load`, and on `sharding_impls.__hash__` — three plausible-looking and entirely wrong culprits.

I filed #40412 before finding this one; closing that as a duplicate.

**Comment by hawkinsp** [comment]
I think we can close this as fixed in JAX 0.11.1? Note 0.11.1 was released August 17.

**Comment by hawkinsp** [comment]
BTW: you are using Python 3.11, for which we dropped support in JAX 0.11. Update to a newer Python to get the updated JAX.
